homeassistant/components/lifx/light.py

```diff
@@ -254,17 +258,21 @@ def merge_hsbk(base, change):
 class LIFXManager:
     """Representation of all known LIFX entities."""
 
-    def __init__(self, hass, platform, async_add_entities):
+    def __init__(self, hass, platform, config_entry, async_add_entities):
         """Initialize the light."""
         self.entities = {}
         self.hass = hass
         self.platform = platform
+        self.config_entry = config_entry
         self.async_add_entities = async_add_entities
         self.effects_conductor = aiolifx_effects().Conductor(hass.loop)
         self.discoveries = []
         self.cleanup_unsub = self.hass.bus.async_listen(
             EVENT_HOMEASSISTANT_STOP, self.cleanup
         )
+        self.entity_registry_updated_unsub = self.hass.bus.async_listen(
```

Comment:
I want to clean up as each entity is deleted, I don't think `async_on_unload` would do that?

Reply:
I'm referring to the unsubscribe callback that we call when cleaning up the manager. Instead of storing that in `entity_registry_updated_unsub` we can pass it to `config_entry.async_on_unload`.

Reply:
@MartinHjelmare I looked into this and it seems that `async_on_unload` is only for deleting a config entry, it is not called on shutdown? I realize that cleanup of this listener is probably not strictly required on shutdown but I would find it odd to have two different cleanup paths.

Reply:
@bdraco For performance and simplcity, I think it would be even better if core would clean up empty devices on its own. Then we would not need the listener (or this PR) at all.

Reply:
We prefer to use this pattern as there's normally less code needed. But it's not a big deal.

Reply:
> @bdraco For performance and simplcity, I think it would be even better if core would clean up empty devices on its own. Then we would not need the listener (or this PR) at all.

I'm trying to think of a case where we would not want to remove a device when it has no entities.

Reply:
For Tuya we create a device but not entities for devices that we know exist but are not supported (yet).

Reply:
Maybe its safe to remove the device when the last entity is removed ?

Reply:
Yeah, but I think it's more common that there are no entities created, besides auto-restored ones, for stale devices after a restart. There's no case of removal of entities that we can trigger on then.

Reply:
Yeah it seems like it would need to be opt-in.

Maybe

`entry.enable_device_cleanup()` ....

Reply:
> There's no case of removal of entities that we can trigger on then.

I don't understand. Users can still remove auto-restored entities and they will – but are then left with lingering, empty devices. Those can be cleaned up automatically which is what this PR does (but only for LIFX).

Yeah, opt in/out would probably be needed to handle all cases but that is preferable to dozens of integrations having to implement something like this PR.

Reply:
Also, 27k+ views say that users would like a UI button to delete a device which I think could remove the guesswork. https://community.home-assistant.io/t/delete-individual-device-and-entity-from-integration/157428

Reply:
I'll try to take a stab at this

Reply:
> I don't understand. Users can still remove auto-restored entities and they will – but are then left with lingering, empty devices.

@amelchio yes, you're correct.

Reply:
https://github.com/home-assistant/core/pull/66209

Reply:
That's great @bdraco.

Since it looks like there is some core movement on this feature I will not implement the post-merge comments for now.